internal/overlord/servstate/handlers.go

```
@@ -51,183 +54,528 @@ var (
 	}
-	cmd := exec.Command(args[0], args[1:]...)
-	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	s.cmd = exec.Command(args[0], args[1:]...)
+	s.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 
-	// Start as another user if specified in plan
-	uid, gid, err := osutil.NormalizeUidGid(service.UserID, service.GroupID, service.User, service.Group)
+	// Start as another user if specified in plan.
+	uid, gid, err := osutil.NormalizeUidGid(s.config.UserID, s.config.GroupID, s.config.User, s.config.Group)
 	if err != nil {
 		return err
 	}
 	if uid != nil && gid != nil {
-		setCmdCredential(cmd, &syscall.Credential{
+		setCmdCredential(s.cmd, &syscall.Credential{
 			Uid: uint32(*uid),
 			Gid: uint32(*gid),
 		})
 	}
 
-	// Pass service description's environment variables to child process
-	cmd.Env = os.Environ()
-	for k, v := range service.Environment {
-		cmd.Env = append(cmd.Env, k+"="+v)
+	// Pass service description's environment variables to child process.
+	s.cmd.Env = os.Environ()
+	for k, v := range s.config.Environment {
+		s.cmd.Env = append(s.cmd.Env, k+"="+v)
 	}
-	logBuffer := servicelog.NewRingBuffer(maxLogBytes)
+
+	// Set up stdout and stderr to write to log ring buffer.
 	var outputIterator servicelog.Iterator
-	if m.serviceOutput != nil {
-		outputIterator = logBuffer.TailIterator()
+	if s.manager.serviceOutput != nil {
+		outputIterator = s.logs.HeadIterator(0)
 	}
+	logWriter := servicelog.NewFormatWriter(s.logs, s.config.Name)
+	s.cmd.Stdout = logWriter
+	s.cmd.Stderr = logWriter
 
-	logWriter := servicelog.NewFormatWriter(logBuffer, req.Name)
-	cmd.Stdout = logWriter
-	cmd.Stderr = logWriter
-	err = cmd.Start()
+	// Start the process!
+	err = s.cmd.Start()
 	if err != nil {
 		if outputIterator != nil {
 			_ = outputIterator.Close()
 		}
-		_ = logBuffer.Close()
+		_ = s.logs.Close()
 		return fmt.Errorf("cannot start service: %v", err)
 	}
+	startTime, _ := s.config.ParseStartTime() // ignore error; it's already been validated
```

The comment explaining why it's okay to ignore the error is longer than the code to check the error. :)

Instead of trusting some remote logic that must remain true for this to not be an awkward issue, we can just move this logic up and return err, simplifying assumptions and reading.

---

True. This code is actually gone now, as we're parsing it to a duration directly in the plan code now.